lib/cloud_controller/paging/sequel_paginator.rb

```diff
@@ -34,20 +34,53 @@ def can_paginate_with_window_function?(dataset)
 
     def paginate_with_window_function(dataset, per_page, page, table_name)
       dataset = dataset.from_self if dataset.opts[:distinct]
-      dataset = if dataset.opts[:graph]
-                  dataset.add_graph_aliases(pagination_total_results: [table_name, :pagination_total_results, Sequel.function(:count).*.over])
-                else
-                  dataset.select_append(Sequel.as(Sequel.function(:count).*.over, :pagination_total_results))
-                end
-      records = dataset.limit(per_page, (page - 1) * per_page).all
+
+      paged_dataset = dataset.limit(per_page, (page - 1) * per_page)
+
+      paged_dataset = if dataset.opts[:graph]
+                        paged_dataset.add_graph_aliases(pagination_total_results: [table_name, :pagination_total_results, Sequel.function(:count).*.over])
+                      elsif from_is_table(dataset)
+                        dataset.join_table(
+                          :inner,
+                          paged_dataset.select(Sequel[table_name][:id].as(:tmp_deferred_id)).
+                              select_append(Sequel.as(Sequel.function(:count).*.over, :pagination_total_results)).
+                              as(:tmp_deferred_table),
+                          Sequel[table_name][:id] => Sequel[:tmp_deferred_table][:tmp_deferred_id]
+                        ).select_append(:pagination_total_results)
+                      else
+                        paged_dataset.select_append(Sequel.as(Sequel.function(:count).*.over, :pagination_total_results))
+                      end
+
+      records = paged_dataset.all
+
       count = records.any? ? records.first[:pagination_total_results] : 0
-      records.each { |x| x.values.delete(:pagination_total_results) }
+      records.each do |x|
+        x.values.delete(:pagination_total_results)
+        x.values.delete(:tmp_deferred_id)
+      end
+      [records, count]
+    end
+
+    def paginate_with_extension(dataset, per_page, page, table_name)
+      paged_dataset = dataset.extension(:pagination).paginate(page, per_page)
+      count = paged_dataset.pagination_record_count
+
+      if from_is_table(dataset)
+        paged_dataset = dataset.join_table(
+          :inner,
+          paged_dataset.select(Sequel[table_name][:id].as(:tmp_deferred_id)).as(:tmp_deferred_table),
+          Sequel[table_name][:id] => Sequel[:tmp_deferred_table][:tmp_deferred_id]
+        )
+      end
+
+      records = paged_dataset.all
+      records.each { |x| x.values.delete(:tmp_deferred_id) }
+
       [records, count]
     end
 
-    def paginate_with_extension(dataset, per_page, page)
-      query = dataset.extension(:pagination).paginate(page, per_page)
-      [query.all, query.pagination_record_count]
+    def from_is_table(dataset)
```

Comment:
Nitpick: ruby convention is to make methods that return booleans end in question marks